File: June-LeetCoding-Challenge-2020/Week3/468-validate-ip-address.py
class Solution:
    def validIPAddress(self, IP: str) -> str:
        if IP.count('.') == 3:
            return self.validIPv4(IP)
        elif IP.count(':') == 7:
            return self.validIPv6(IP)
        else:
            return "Neither"
        
    def validIPv4(self, ip4: str) -> str:
        ip_split = ip4.split('.')
        
        for i in ip_split:
            if len(i) == 0 or len(i) > 3: # IPv4 address between 1 and 3
                return "Neither"
            if i[0] == '0' and len(i) != 1 or not i.isdigit() or int(i) > 255:
                return "Neither"
        return "IPv4"
    
    def validIPv6(self, ip6: str) -> str:
        ip_split = ip6.split(':')
        hex_digit = "0123456789abcdefABCDEF"
        for i in ip_split:
            if len(i) == 0 or len(i) > 4 or not all(c in hexdigits for c in i):
                return "Neither"
        return "IPv6"
    # TC: O(N), n is the length of ip string
    # SC: O(1)
    # Runtime: 20 ms, faster than 98.12%
    # Memory Usage: 13.9 MB, less than 41.09%

File: June-LeetCoding-Challenge-2020/Week3/468-validate-ip-address.cpp
// Time Complexity: O(n) ?  CONSTANT?!!
// Space Complexity: O(1) ? CONSTANT?!!
// Runtime: 4 ms, faster than 37.88%
// Memory Usage: 6.3 MB, less than 80.55%
class Solution {
public:
    string IPv4 = "IPv4";
    string IPv6 = "IPv6";
    string neither = "Neither";

    string validIPAddress(string IP) {
        
        if(count(IP.begin(), IP.end(), '.') == 3){
            return validateIPv4(IP);
            
        }else if(count(IP.begin(), IP.end(), ':') == 7){
            return validateIPv6(IP);
            
        }else{
            return neither;
        }
        
    }

    vector<string> split(string s, char c){
        vector<string> parts;
        string part;
        
        istringstream in(s);
        while(getline(in, part, c)){
            parts.push_back(part);
        }
        
        return parts;
    }
    
    string validateIPv4(string IP){
        vector<string> nums = split(IP, '.');
        
        if(nums.size() != 4)
            return neither;
        
        for(string num : nums){
            if(num.empty() || num.size() > 3 || (num.size() >= 2 && num[0] == '0'))
                return neither;
            
            for(char n : num) {
                if(!isdigit(n)){
                    return neither;
                }
            }
            
            if(stoi(num) > 255)
                return neither;
            
        }
        
        return IPv4;
    }
    
    string validateIPv6(string IP){
        vector<string> nums = split(IP, ':');
        
        if(nums.size() != 8)
            return neither;
        
        for(string num : nums){
            if(num.empty() || num.size() > 4)
                return neither;
            
            for(char n : num)
                if( (!isdigit(n) && (!isalpha(n))) || toupper(n) > 'F' )
                    return neither;
            
        }
        
        return IPv6;
    }
};